elide-model-config/src/main/java/com/yahoo/elide/modelconfig/store/ConfigDataStoreTransaction.java

```diff
@@ -0,0 +1,211 @@
+import com.yahoo.elide.modelconfig.validator.Validator;
+
+import org.apache.commons.io.FileUtils;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Elide DataStoreTransaction which loads/persists HJSON configuration files as Elide models.
+ */
+public class ConfigDataStoreTransaction implements DataStoreTransaction {
+
+
+    private final FileLoader fileLoader;
+    private final Set<Runnable> todo;
+    private final Set<ConfigFile> dirty;
+    private final Set<String> deleted;
+    private final Validator validator;
+    private final boolean readOnly;
+
+    public ConfigDataStoreTransaction(
+            FileLoader fileLoader,
+            boolean readOnly,
+            Validator validator
+    ) {
+        this.fileLoader = fileLoader;
+        this.readOnly = readOnly || !fileLoader.isWriteable();
+        this.dirty = new LinkedHashSet<>();
+        this.deleted = new LinkedHashSet<>();
+        this.todo = new LinkedHashSet<>();
+        this.validator = validator;
+    }
+
+    @Override
+    public <T> void save(T entity, RequestScope scope) {
+        if (readOnly) {
+            throw new UnsupportedOperationException("Configuration is read only.");
+        }
+
+        ConfigFile file = (ConfigFile) entity;
+        dirty.add(file);
+        todo.add(() -> updateFile(file.getPath(), file.getContent()));
+    }
+
+    @Override
+    public <T> void delete(T entity, RequestScope scope) {
+        if (readOnly) {
+            throw new UnsupportedOperationException("Configuration is read only.");
+        }
+        ConfigFile file = (ConfigFile) entity;
+        dirty.add(file);
+        deleted.add(file.getPath());
+        todo.add(() -> deleteFile(file.getPath()));
+    }
+
+    @Override
+    public void flush(RequestScope scope) {
+        if (!readOnly) {
+            Map<String, ConfigFile> resources;
+            try {
+                resources = fileLoader.loadResources();
+            } catch (IOException e) {
+                throw new IllegalStateException(e);
+            }
+
+            for (ConfigFile file : dirty) {
+                resources.put(file.getPath(), file);
+            }
+
+            for (String path: deleted) {
+                resources.remove(path);
+            }
+
+            try {
+                validator.validate(resources);
+            } catch (Exception e) {
+                throw new BadRequestException(e.getMessage());
+            }
+        }
+    }
+
+    @Override
+    public void commit(RequestScope scope) {
+        boolean validateOnly = scope.getRequestHeaderByName(VALIDATE_ONLY_HEADER) != null;
+
+        if (! validateOnly) {
+            for (Runnable runnable : todo) {
+                runnable.run();
```

Comment:
Do we need some rollback mechanism here. For example while saving 5 files, it was able to save only 2 and 3rd failed. Will it leave the Agg store in a bad state.

Reply:
I thought about this but took a different route.  There are two kinds of errors - recoverable and unrecoverable.  Unrecoverable would be things like disk corruption or file system being full.  Rollback is not likely to succeed in these cases.  Recoverable would be things like permission errors.

Instead of implementing rollback (which could be messy/complicated), I added extra validation around permission errors.